content/blog/2023-03-23-when-a-date-breaks-booting-the-kernel.md

```diff
@@ -0,0 +1,213 @@
+Title: When a date breaks booting the kernel
+Date: 2023-03-23 00:00:00
+Authors: Benjamin Gray
+Category: Development
+Tags: linux
+
+## The setup
+
+I've recently been working on internal CI infrastructure for testing kernels before sending them to the mailing list. As part of this effort, I became interested in reproducible builds. Minimising the changing parts outside of the source tree itself could improve consistency and ccache hits, which is great for trying to make the CI faster and more reproducible.
+
+As part of this effort, I came across the `KBUILD_BUILD_TIMESTAMP` environment variable. This variable is used to set the kernel timestamp, which is primarily just baked into the version string for any users who want to know when their kernel was built. That's mostly irrelevant for our work, so an easy `KBUILD_BUILD_TIMESTAMP=0` later and... it still uses the current date.
+
+Ok, after checking the documentation it looks like the timestamp variable is actually expected to be a date format, as accepted by `date -d`. So to make it obvious that it's not a 'real' date, let's set `KBUILD_BUILD_TIMESTAMP=0000-01-01`.
+
+Building and booting the kernel, we see `#1 SMP 0000-01-01` printed as the build timestamp. Success! After confirming everything works, I set the environment variable in the CI jobs and call it a day.
+
+A few days later I need to run the CI to test my patches, and something strange happens. It builds fine, but the boot tests that load a root disk image fail inexplicably: there is a kernel panic saying "VFS: Unable to mount root fs on unknown-block(253,2)".
```

Comment:
- the kernel log claimed vda2 didn't exist, and then printed out a block device list that very much did include vda2
- you don't mention the /dev/console error
- the -2 (ENOENT) errno was also a helpful pointer in terms of working out that it was due to /dev not existing